File: src/Host/ClientApp/src/entities/Playthrough.ts
import type { BaseQueryFn, FetchArgs, FetchBaseQueryError, FetchBaseQueryMeta } from '@reduxjs/toolkit/query';
import type { EndpointBuilder } from '@reduxjs/toolkit/query';
import { apiRequester, type TagType } from "../shared/ApiRequester"
import type { ApiMeta } from './ApiMeta';

export interface PlaythroughState {
    data: Playthrough,
    isLoading: boolean,
    isChecked: boolean,
    error: string
}

export interface Playthrough {
    gameSessionId: number,
    currentFragmentId: number,
    chapterNumber: number,
    title: string,
    slides: Slide[],
    currentSlideIndex: number,
    choices: StoryChoice[]
}

export interface Slide {
    id: number
    text: string[]
    imageName: string
}

export interface StoryChoice {
    fragmentId: number
    text: string
}

export const createCurrentStoryMutation = <BodyType extends Record<string, unknown>>(
    url: string,
    builder: EndpointBuilder<BaseQueryFn<string | FetchArgs, unknown, FetchBaseQueryError, ApiMeta, FetchBaseQueryMeta>, TagType, "apiRequester">
) => {
    return builder.mutation<Playthrough, BodyType>({
        query: (body) => ({
            url,
            method: 'POST',
            body,
        }),
        async onQueryStarted(_, { dispatch, queryFulfilled }) {
            const { data } = await queryFulfilled;
            dispatch(
                extendedApiSlice.util.upsertQueryData('getPlaythrough', undefined, data)
            );
        },
        invalidatesTags:['StoryList', 'Story']
    });
};

const extendedApiSlice = apiRequester.injectEndpoints({
    endpoints: (builder) => ({

        getPlaythrough: builder.query<Playthrough, void>({
            query: () => 'playthrough/getPlaythrough',
            providesTags: ['Playthrough'],
        }),

        setChoice: createCurrentStoryMutation<{
            storyNodeId: number;
            choiceNumber: number;
        }>('/playthrough/setChoice', builder),

        dropPlaythrough: createCurrentStoryMutation('/playthrough/dropPlaythrough', builder),
    }),
});


export const {
    useGetPlaythroughQuery,
    useSetChoiceMutation,
    useDropPlaythroughMutation
} = extendedApiSlice;

File: src/Domain/Story/Playthrough.cs
using YAGO.World.Domain.Slides;

namespace YAGO.World.Domain.Story
{
    public class Playthrough
    {
        public long GameSessionId { get; }
        public long CurrentFragmentId { get; }
        public int ChapterNumber { get; }
        public string Title { get; }
        public Slide[] Slides { get; }
        public int CurrentSlideIndex { get; }
        public StoryChoice[] Choices { get; }

        public Playthrough(
            long gameSessionId,
            long currentFragmentId,
            int chapterNumber,
            string title,
            Slide[] slides,
            int currentSlideIndex,
            StoryChoice[] choices)
        {
            GameSessionId = gameSessionId;
            CurrentFragmentId = currentFragmentId;
            ChapterNumber = chapterNumber;
            Title = title;
            Slides = slides;
            CurrentSlideIndex = currentSlideIndex;
            Choices = choices;
        }
    }
}
